File: tests/reboot/std/collections/queue/v1/queue_tests.ts
// `eslint` disabled to preserve separation of imports for documentation.
import { Value } from "@bufbuild/protobuf";
import { Application, Reboot } from "@reboot-dev/reboot";
import { Queue } from "@reboot-dev/reboot-std/collections/queue/v1";
// eslint-disable-next-line
import { queueLibrary } from "@reboot-dev/reboot-std/collections/queue/v1";
import { sortedMapLibrary } from "@reboot-dev/reboot-std/collections/v1/sorted_map.js";
import { strict as assert } from "node:assert";
import { test } from "node:test";
import { CreateRequest } from "../../../../greeter_rbt.js";

test("Use queue servicers", async (t) => {
  // These tests simply verify that we are properly exporting the Queue
  // servicers to be used in typescript. They aren't meant to
  // thoroughly test any logic.

  let rbt: Reboot;
  t.beforeEach(async () => {
    rbt = new Reboot();
    await rbt.start();
  });
  t.afterEach(async () => {
    await rbt.stop();
  });

  await t.test("Interact with the Queue", async (t) => {
    await rbt.up(
      new Application({ libraries: [queueLibrary(), sortedMapLibrary()] })
    );

    const context = rbt.createExternalContext("test", {
      appInternal: true,
    });

    const myQueue = await Queue.ref("test-id");
    await myQueue.enqueue(context, {
      items: [
        { value: Value.fromJson(1) },
        { value: Value.fromJson(3) },
        { value: Value.fromJson(303) },
      ],
    });

    let bulkItems = await myQueue.dequeue(context, { bulk: true, atMost: 2 });

    assert(bulkItems.items.length === 2, `Expected 2 items to be dequeued.`);
    assert(
      bulkItems.items[0].value?.toJson() === 1,
      `Expected first dequeued item to be 1, but it was not.`
    );
    assert(
      bulkItems.items[1].value?.toJson() === 3,
      `Expected second dequeued item to be 3, but it was not.`
    );

    let lastItem = await myQueue.dequeue(context);

    assert(
      lastItem.value?.toJson() === 303,
      `Expected last dequeued item to be 303, but it was not.`
    );
  });

  await t.test(
    "Examples for documentation for single enqueue/dequeue.",
    async (t) => {
      await rbt.up(
        new Application({ libraries: [queueLibrary(), sortedMapLibrary()] })
      );

      const context = rbt.createExternalContext("test", {
        appInternal: true,
      });

      const firstQueue = Queue.ref("my-first-queue");
      const secondQueue = Queue.ref("my-second-queue");

      await firstQueue.enqueue(context, {
        value: Value.fromJson({ details: "details-go-here" }),
      });

      await secondQueue.enqueue(context, {
        bytes: new TextEncoder().encode("my-bytes"),
      });

      // `Any` not supported by TypeScript because conversion from TS to Python is
      // passed via JSON, which bufbuilder needs a registry for. Converting into
      // bytes instead would fix this.
      // await thirdQueue.enqueue(context, {
      //   any: Any.pack(any),
      // });

      const { value } = await firstQueue.dequeue(context);
      console.log(value?.toJson()["details"]);

      const { bytes } = await secondQueue.dequeue(context);
      console.log(bytes);

      // `Any` not supported by TypeScript because conversion from TS to Python is
      // passed via JSON, which bufbuilder needs a registry for. Converting into
      // bytes instead would fix this.
      // const { any } = await thirdQueue.dequeue(context);
      // console.log(any);
    }
  );

  await t.test(
    "Examples for documentation for bulk enqueue/dequeue.",
    async (t) => {
      await rbt.up(
        new Application({ libraries: [queueLibrary(), sortedMapLibrary()] })
      );

      const context = rbt.createExternalContext("test", {
        appInternal: true,
      });

      const firstQueue = Queue.ref("my-first-queue");
      const secondQueue = Queue.ref("my-second-queue");

      // `any` needs to be defined for documentation. It won't actually be
      // shown. Any proto will do.
      let any = new CreateRequest({
        title: "king",
        name: "nemo",
        adjective: "fishy",
      });

      await firstQueue.enqueue(context, {
        items: [
          { value: Value.fromJson(null) },
          { value: Value.fromJson(true) },
          { value: Value.fromJson(3) },
          { value: Value.fromJson("apple") },
          { value: Value.fromJson(["a", "b", "c"]) },
          { value: Value.fromJson({ details: "details-go-here" }) },
        ],
      });

      await secondQueue.enqueue(context, {
        items: [
          { bytes: new TextEncoder().encode("some-bytes") },
          { bytes: new TextEncoder().encode("some-more-bytes") },
        ],
      });

      // `Any` not supported by TypeScript because conversion from TS to Python is
      // passed via JSON, which bufbuilder needs a registry for. Converting into
      // bytes instead would fix this.
      // await thirdQueue.enqueue(context, { items: [{ any }, { any }] });

      const { items } = await firstQueue.dequeue(context, {
        bulk: true,
        atMost: 5,
      });
      // `items` is a list of Items
    }
  );

  await t.test("Test tryDequeue()", async (t) => {
    await rbt.up(
      new Application({ libraries: [queueLibrary(), sortedMapLibrary()] })
    );

    const context = rbt.createExternalContext("test", {
      appInternal: true,
    });

    const queue = Queue.ref("test-queue");

    // `console.log()` and if/else below is pulled in for example code in documentation.
    const response = await queue.tryDequeue(context, { bulk: true, atMost: 1 });
    if (response.items.length === 0) {
      console.log("The queue is empty.");
    } else {
      // Something was in the queue; handle it.
    }

    assert(response.items.length === 0, `Expected 0 items to be dequeued.`);

    // Add item and `tryDequeue()` it.
    await queue.enqueue(context, {
      value: Value.fromJson({ details: "details-go-here" }),
    });

    const { value } = await queue.tryDequeue(context);

    assert(
      value?.toJson()["details"] === "details-go-here",
      `Expected to tryDequeue() the correct item.`
    );
  });

  await t.test("Test empty().", async (t) => {
    await rbt.up(
      new Application({ libraries: [queueLibrary(), sortedMapLibrary()] })
    );

    const context = rbt.createExternalContext("test", {
      appInternal: true,
    });

    const queue = Queue.ref("test-queue");

    // Need to construct state before reading.
    await queue.tryDequeue(context);

    // `console.log()` below is pulled in for example code in documentation.
    const response = await queue.empty(context);
    if (response.empty) {
      console.log("The queue is empty.");
    }

    assert(
      response.empty === true,
      `Expected empty() to be true when there are no items.`
    );

    // Add item and check `empty()` again.
    await queue.enqueue(context, {
      value: Value.fromJson({ details: "details-go-here" }),
    });

    const emptyFalseResponse = await queue.empty(context);

    assert(
      emptyFalseResponse.empty === false,
      `Expected empty() to be true when there are items.`
    );
  });
});

File: tests/reboot/std/collections/queue/v1/queue_tests.py
import asyncio
import unittest
from reboot.aio.applications import Application
from reboot.aio.tests import Reboot
from reboot.protobuf import as_int, as_str, from_int, from_str, pack, unpack
from reboot.std.collections.queue.v1.queue import (
    DEFAULT_BULK_COUNT,
    DequeueResponse,
    Queue,
    servicers,
)
from reboot.std.item.v1.item import Item
from tests.reboot.greeter_rbt import CreateRequest


class TestQueue(unittest.IsolatedAsyncioTestCase):

    async def asyncSetUp(self) -> None:
        self.rbt = Reboot()
        await self.rbt.start()

    async def asyncTearDown(self) -> None:
        await self.rbt.stop()

    async def test_single_value(self) -> None:
        """
        Test that we can do a simple enqueue and a dequeue with a `Value`.
        """
        await self.rbt.up(Application(
            servicers=servicers(),
        ))

        context = self.rbt.create_external_context(
            name=f"test-{self.id()}",
            app_internal=True,
        )

        queue = Queue.ref("test-queue")

        # Enqueue single `Value`.
        await queue.enqueue(context, value=from_str("tree"))

        # Dequeue single `Value`.
        item = await queue.dequeue(context)

        # Test which field is set.
        self.assertFalse(item.HasField("any"))
        self.assertFalse(item.HasField("bytes"))
        self.assertTrue(item.HasField("value"))

        # Test expected values.
        self.assertEqual(as_str(item.value), "tree")

    async def test_single_bytes(self) -> None:
        """
        Test that we can do a simple enqueue and a dequeue with a `bytes`.
        """
        await self.rbt.up(Application(
            servicers=servicers(),
        ))

        context = self.rbt.create_external_context(
            name=f"test-{self.id()}",
            app_internal=True,
        )

        queue = Queue.ref("test-queue")

        # Enqueue single `bytes`.
        await queue.enqueue(context, bytes=b"brownies")

        # Dequeue single `bytes`.
        item = await queue.dequeue(context)

        # Test which field is set.
        self.assertFalse(item.HasField("any"))
        self.assertTrue(item.HasField("bytes"))
        self.assertFalse(item.HasField("value"))

        # Test expected values.
        self.assertEqual(item.bytes, b"brownies")

    async def test_single_any(self) -> None:
        await self.rbt.up(Application(
            servicers=servicers(),
        ))

        context = self.rbt.create_external_context(
            name=f"test-{self.id()}",
            app_internal=True,
        )

        queue = Queue.ref("test-queue")

        # Enqueue single `Any`.
        # Any proto will work here.
        message = CreateRequest(
            title="king",
            name="nemo",
            adjective="fishy",
        )
        await queue.enqueue(context, any=pack(message))

        # Dequeue single `Any`.
        item = await queue.dequeue(context)
        unpacked_item = unpack(item.any, CreateRequest)

        self.assertEqual(unpacked_item.title, "king")
        self.assertEqual(unpacked_item.name, "nemo")
        self.assertEqual(unpacked_item.adjective, "fishy")

    async def test_bulk_enqueue_items(self) -> None:
        """
        Test that we can do a bulk enqueue and a dequeue.
        """
        await self.rbt.up(Application(
            servicers=servicers(),
        ))

        context = self.rbt.create_external_context(
            name=f"test-{self.id()}",
            app_internal=True,
        )

        queue = Queue.ref("test-queue")

        # Enqueue multiple `Item`s.
        await queue.enqueue(
            context, items=[
                Item(value=from_str("apple")),
                Item(value=from_str("gum")),
                Item(value=from_str("ice")),
            ]
        )
        await queue.enqueue(
            context, items=[
                Item(value=from_str("watch")),
                Item(value=from_str("lemon")),
            ]
        )

        # Dequeue two.
        bulk_items = await queue.dequeue(context, bulk=True, at_most=2)
        self.assertEqual(len(bulk_items.items), 2)
        self.assertEqual(as_str(bulk_items.items[0].value), "apple")
        self.assertEqual(as_str(bulk_items.items[1].value), "gum")

        # Dequeue two.
        bulk_items = await queue.dequeue(context, bulk=True, at_most=2)
        self.assertEqual(len(bulk_items.items), 2)
        self.assertEqual(as_str(bulk_items.items[0].value), "ice")
        self.assertEqual(as_str(bulk_items.items[1].value), "watch")

    async def test_dequeue_more_items_than_enqueued(self) -> None:
        """
        Test that we can dequeue fewer items than `at_most`.
        """
        await self.rbt.up(Application(
            servicers=servicers(),
        ))

        context = self.rbt.create_external_context(
            name=f"test-{self.id()}",
            app_internal=True,
        )

        queue = Queue.ref("test-queue")

        # Enqueue multiple `Item`s.
        await queue.enqueue(
            context, items=[
                Item(value=from_str("apple")),
                Item(value=from_str("gum")),
            ]
        )

        # Try to dequeue three.
        bulk_items = await queue.dequeue(context, bulk=True, at_most=3)
        self.assertEqual(len(bulk_items.items), 2)
        self.assertEqual(as_str(bulk_items.items[0].value), "apple")
        self.assertEqual(as_str(bulk_items.items[1].value), "gum")

    async def test_try_dequeue_before_enqueue(self) -> None:
        """
        Test that we can dequeue (non-blocking) before we enqueue.
        """
        await self.rbt.up(Application(
            servicers=servicers(),
        ))

        context = self.rbt.create_external_context(
            name=f"test-{self.id()}",
            app_internal=True,
        )

        queue = Queue.ref("test-queue")

        # Attempt to dequeue.
        response = await queue.try_dequeue(context, bulk=True)
        self.assertEquals(response, DequeueResponse())

    async def test_dequeue_bulk_default_at_most(self) -> None:
        """
        Test that when we bulk dequeue, a default value gets set for `at_most`.
        """
        await self.rbt.up(Application(
            servicers=servicers(),
        ))

        context = self.rbt.create_external_context(
            name=f"test-{self.id()}",
            app_internal=True,
        )

        queue = Queue.ref("test-queue")

        # Enqueue twice the default number of items to dequeue.
        items = [
            Item(value=from_int(n)) for n in range(DEFAULT_BULK_COUNT * 2)
        ]
        await queue.enqueue(context, items=items)

        response = await queue.dequeue(context, bulk=True)
        self.assertEqual(
            len(response.items), DEFAULT_BULK_COUNT,
            "Get a default number of items back when not setting `at_most`"
        )
        for i in range(DEFAULT_BULK_COUNT):
            self.assertEqual(as_int(response.items[i].value), i)

    async def test_dequeue_waits_for_item(self) -> None:
        await self.rbt.up(Application(
            servicers=servicers(),
        ))
        queue_name = "test-queue"
        has_dequeued = False

        async def dequeue():
            nonlocal has_dequeued
            context = self.rbt.create_external_context(
                name=f"test-{self.id()}-dequeue",
                app_internal=True,
            )
            response = await Queue.ref(queue_name).dequeue(
                context,
                bulk=True,
                at_most=10,
            )
            self.assertEqual(len(response.items), 3)
            self.assertEqual(as_str(response.items[0].value), "apple")
            self.assertEqual(as_str(response.items[1].value), "gum")
            self.assertEqual(as_str(response.items[2].value), "ice")
            has_dequeued = True

        dequeue_task: asyncio.Task = asyncio.create_task(dequeue())

        # Check that the dequeue is still waiting.
        self.assertFalse(
            has_dequeued,
            "Should not have managed to dequeue items before there are items."
        )

        # Enqueue items.
        context = self.rbt.create_external_context(
            name=f"test-{self.id()}-enqueue",
            app_internal=True,
        )
        await Queue.ref(queue_name).enqueue(
            context, items=[
                Item(value=from_str("apple")),
                Item(value=from_str("gum")),
                Item(value=from_str("ice")),
            ]
        )

        # Wait for the dequeue to finish.
        await dequeue_task

        self.assertTrue(
            has_dequeued, "Should have dequeued after items were enqueued."
        )


if __name__ == '__main__':
    unittest.main()
